[#1, #2, #3, #4] feature: member관련 CRUD, 로그인, 로그아웃 및 관련 테스트 코드 작성

jongwanra opened the pull request:
##  *Self CheckList*

- [x] 한 메서드에 오직 한 단계의 들여쓰기하기
- [x] else keyword 사용하지 않기
- [x] 변수명, 메서드명 줄여쓰지 않기

## *[Optional] ETC*

* #1 
* #2
* #3 
* #4

jongwanra committed:
[#1] feature: Member Entity 추가

jongwanra committed:
[#1] feature: 공통 사용 BaseEntity 추가 및 Auditing 적용

jongwanra committed:
[#1] feature: MemberRepository, MemberService 추가

jongwanra committed:
[#1] test: MemberService SignUpTest 추가

jongwanra committed:
[#1] feature: SecurityConfig 및 PasswordEncoder 추가

jongwanra committed:
[#2] feature: GlobalExceptionHandler & Common Response, ErrorResponse 추가

jongwanra committed:
[#1] feature: MemberSignUpRequestDto Validation 로직 추가

jongwanra committed:
[#1] feature: nickname, loginId 중복 여부 체크 서비스 로직 추가

jongwanra committed:
[#1] feature: 인가 관련 LoginMember, LoginMemberArgumentResolver 추가

jongwanra committed:
[#1] feature: Auth domain 추가

jongwanra committed:
[None] chore: wrapper class -> primitive type으로 변경 (Entity 제외)

jongwanra committed:
[None] test: MockMVC 테스트 코드 Member, Auth 전체 추가 및 RestDocs  적용

jongwanra committed:
[None] test: MockMVC 테스트 코드 Member, Auth 전체 추가 및 RestDocs  적용

jongwanra committed:
[None] chor: UriPrefix 한 곳에서 관리하는 UriPrefix 추가

jongwanra committed:
[None] chore: Member @Column Annotation @Id 제외하고 삭제

jongwanra committed:
[None] chore: Script 기반 DDL 적용

jongwanra committed:
[#4] feature: CI Test 배포(1)

jongwanra committed:
[#4] feature: CI Test 배포(2)

jongwanra committed:
[#4] feature: CI Test 배포(3)

jongwanra committed:
[None] chore: annotation들에 대한 설명 작성

jongwanra committed:
[None] chore: test 환경 annotation 섦명 추가

jongwanra committed:
[None] chore: 빌드 시, restdocs 업데이트 로직 추가

jongwanra committed:
[None] chore: 빌드 파일로 애플리케이션 실행 시 restdocs 보이지 않는 이슈 해결

jongwanra committed:
[None] chore: RestDocs 파일 형식 변경 및 통합

jongwanra committed:
[None] feat: JpaAuditingConfig에 HttpSession 적용

jongwanra committed:
[None] feat: JpaAuditingConfig에 HttpSession 적용

f-lab-jd commented on build.gradle:
```diff
@@ -1,35 +1,86 @@
-	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
리뷰드렸던것처럼 security 의존이 꼭 필요한지 다시 한번 고민해보시면 좋을 것 같습니다~

jongwanra replied:
넵! JD님께서 말씀해주신 부분 대로 단순히 암호화 만을 위해서 하나의 프레임 워크인 spring-security의 의존성을 갖는 다는 것은 오버 엔지니어링 이라고 저도 공감을 할 수 있었습니다!
build.gradle 에서 제거한 이후에 Password Encoder는  따로 java.security.MessageDigest 패키지를 활용해서 구현했습니다!

f-lab-jd commented on build.gradle:
```diff
@@ -1,35 +1,86 @@
+
+asciidoctor.doFirst {
+    delete file('src/main/resources/static/docs')
+}
```
다음 멘토링 시간에 restdocs 를 만드는 과정에 대해서도 한번 설명해주시면 좋을 것 같습니다.

f-lab-jd commented on src/main/java/study/outfitoftheday/OutfitOfTheDayApplication.java:
```diff
@@ -2,8 +2,22 @@
+ *
+ * 아래의 annotation들을 포함한다.
+ * 1. @SpringBootConfiguration: Spring의 Bean 구성 클래스임을 지정한다.
+ * 2. @EnableAutoConfiguration: Spring Boot의 자동 구성 기능을 활성화한다. 클래스 패스와 설정에 따라 애플리케이션을 설정하고 초기화하는 데 필요한 빈들을 자동으로 추가한다.
```
autoconfigure 는 내부적으로 어떻게 동작할까요~?

jongwanra replied:
AutoConfigure은 내부적으로 아래와 같이 동작합니다!

1. @SpringBootApplication 
2. @EnableAutoConfiguration
3. @Import(AutoconfigurationImportSelector.class) 
4. META-INF/spring/org.springframework.boot.autoconfigure.AutoConfiguration.imports 해당 파일을 열어서 설정 정보들을 가져옵니다. 
5. 이후에 해당 파일의 설정 정보가 스프링 컨테이너에 등록되고 사용됩니다.

해당 경로에서 자동 설정 되는 목록들 중에 
JdbcTemplateAutoConfiguration, DataSourceTransactionManagerAutoConfiguration, DataSourceAutoConfiguration
Spring Boot가 구동 시 자동으로 라이브러리들을  배치해줘서 저희가 편리하게 개발을 할 수 있게 되는 거였네요 ㅎㅎ

f-lab-jd commented on src/main/java/study/outfitoftheday/common/config/JpaAuditingConfig.java:
```diff
@@ -0,0 +1,46 @@
+
+	@Bean
+	public AuditorAware<Long> auditorProvider() {
+		System.out.println("httpSession = " + httpSession);
```
`System.out.println` 은 실제 운영하는 서비스에서는 절대 사용하지 않는다고 말씀드린적이 있었는데 혹시 왜인지 기억하시나요~?

jongwanra replied:
네넵!  

System.out.println() method를 사용한다는 것은 표준 출력 스트림(stdout)에 데이터를 출력하는 것을 의미하는데요, 
이 작업은 kernel mode에서 발생합니다.  Blocking I/O 방식으로 동작하기 때문에  데이터 출력이 완료될 때까지 실행 중이던 Thread가 블로킹 되게 됩니다. 그렇기 때문에 사용해서는 안된다고 기억이 납니다! 

해당 method는 바로 삭제 진행하고 System.out.println() method 대신 logging을 무조건 적으로 사용하는 것으로 
컨벤션에도 적어 두겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/common/config/JpaAuditingConfig.java:
```diff
@@ -0,0 +1,46 @@
+	@Bean
+	public AuditorAware<Long> auditorProvider() {
+		System.out.println("httpSession = " + httpSession);
+		if (httpSession == null) {
```
httpSession 이 null 인 경우가 있을까요?

jongwanra replied:
HttpSession의 생성시점에 대해서 다시 살펴보게 되었습니다🤔
null인 경우는 없습니다.  HttpSesison을 외부에서 주입하는 방식으로 적용했을 때 기본적으로 Proxy 객체로 생성해주기 때문입니다!

이 부분도 수정해두겠습니다🙏

f-lab-jd commented on src/main/java/study/outfitoftheday/common/config/JpaAuditingConfig.java:
```diff
@@ -0,0 +1,46 @@
+		if (httpSession == null) {
+			return () -> Optional.empty();
+		}
+		return () -> Optional.ofNullable((Long)httpSession.getAttribute("MEMBER_ID"));
```
매직넘버, 스트링은 상수로 빼는 것이 좋습니다. 왜일까요~?

jongwanra replied:
상수로 빼는 것이 좋은 이유는, 

 **무엇을 의미하는지 확신할 수 없기 때문입니다.**

상수로 선언함으로써 불분명한 값들에 의미를 부여하고 어떤 역할을 할 수 있는지 명확하게 함으로서 매직넘버, 스트링에 의미를 파악하는데 드는 비용을 아낄 수 있습니다 ㅎㅎ

상수로 뺴는 작업 진행하도록 하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/auth/exception/NoAccessAuthorizationException.java:
```diff
@@ -0,0 +1,7 @@
+package study.outfitoftheday.core.domain.auth.exception;
+
+public class NoAccessAuthorizationException extends RuntimeException {
+	public NoAccessAuthorizationException() {
```
기본 생성자이외에도 메시지를 포함시키는도  좋지 않을까요~?

jongwanra replied:
넵! 기본 생성자 이외에  메시지를 포함시키는 로직으로 변경 했습니다! 
추가적으로 멘토링 시간에 말씀하셨던 부분 중에 공통으로 Custom Exception을 관리하는 방법으로 생각하여 RuntimeException을 
확장한 ServiceException을 만들었고 해당 ServiceException을 확장해서 Custom Exception을 생성하려고 합니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/common/interceptor/RequiredAuthInterceptor.java:
```diff
@@ -0,0 +1,40 @@
+	private final AuthService authService;
+
+	@Override
+	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws
```
`throws Exception` 이 반드시 필요한가요~?

jongwanra replied:
아닙니다! 넣지 않아도 되는 키워드로 확인됩니다. 삭제하겠습니다!🙏

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/auth/service/AuthService.java:
```diff
@@ -0,0 +1,13 @@
+
+import study.outfitoftheday.core.domain.member.entity.Member;
+
+public interface AuthService {
```
interface 사용 기준에 대해 다시 한번 고민해보심 좋을 것 같습니다.

jongwanra replied:
`앞으로 확장이 될 수 있는 클래스인가?` 에 대한 기준으로 바라봤을 때, 
 AuthService, MemberService는 그렇지 않다고 생각이 듭니다.

수정해서 반영해 두겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/common/resolver/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
```
이 부분도 `throws Exception` 이 반드시 필요한지 확인해주시면 좋을 것 같고 메소드의 리턴타입이 꼭 Object 일 필요가 있을까요~?

jongwanra replied:
* throws Exception은 불필요한 곳은 전부 체크해서 제거했습니다 ㅎㅎ 
* Return Type은 Object가 아닌 Member로 명확하게 반환 가능해서 수정했습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/auth/service/AuthServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+ * */
+@Service
+@RequiredArgsConstructor
+@Transactional
```
`@Transactional` 어노테이션 클래스에 붙어있으면 어떤일이 벌어질까요? 
현재 클래스에 필요한지 한번 다시 고민해보시면 좋을 것 같습니다.

jongwanra replied:
@Transactional 어노테이션을 붙이면 
스프링은 자동으로  transaction을 시작하고 성공 시 commit(), 예외 발생 시 rollback() method를 호출합니다.

현재 AuthService의 method를 보았을 때 단순하게 자신의 데이터만 조회하는 로직을 수행하고  있기 때문에 
Transaction 없이도 데이터의 일관성을 보장할 수 있다고 여겨집니다! 
해당 내용 반영하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/auth/service/AuthServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+@Service
+@RequiredArgsConstructor
+@Transactional
+@Slf4j
```
`@Slf4j` 어노테이션은 있지만 로그는 없어보입니다.

jongwanra replied:
로그를 사용할 하지 않는 경우에는 @Slf4j 어노테이션을 삭제하는 습관을 가지겠습니다.!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/member/exception/MismatchPasswordInLoginException.java:
```diff
@@ -0,0 +1,9 @@
+package study.outfitoftheday.core.domain.member.exception;
+
+public class MismatchPasswordInLoginException
```
`MismatchPasswordInSignUpException` 클래스와와 구분해서 사용하시는 이유가 있으신가요~?

jongwanra replied:
로그인, 회원가입 시에 비밀번호가 일치하지 않을 경우 전달하는 message 내용이 다를 경우를 생각해서 구분하게 되었습니다! 🤔

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/common/BaseEntity.java:
```diff
@@ -0,0 +1,64 @@
+	 * */
+	@CreatedBy
+	@Column(updatable = false)
+	private Long createdBy;
```
`createdBy,` `updatedBy` 와 매핑되는 컬럼을 Long 타입으로 관리하실 생각이신가요~? 
이름으로 관리하는것에 비해 어떤 장단점이 있을까요?

jongwanra replied:
createdBy, updatedBy를 Long타입으로 관리했을 때  이름으로 관리하는 것에 비해 장점은 떠오르지 않고  단점으로 생각되는 부분들이 생각나네요.

* 처음 회원 가입을 시도 했을 경우 auto_increment 방식으로 PK를 생성하기 때문에  createdBy와 updatedBy에 기본값으로 데이터를 넣을 수 밖에 없었습니다.(null 값을 허용하지 않기 위해서 0L을 기본 값으로 지정했습니다.) 
그에 비해 nickname으로 관리한다면 이러한 부분은 신경쓸 필요가 없이 데이터의 정합성을 유지할 수 있다는 장점이 있습니다.

*  nickname의 값은 member_id의 값보다 데이터를 조인하지 않아도 직관적으로 nickname으로 해당 유저를 알 수 있습니다.  
* 보통 회사 내에서 데이터를 분석하기 위해 사용하는  컬럼으로 쓰이기 때문에 nickname이 더욱 유용한 의미를 지닐 수 있습니다. 
* post, comment table 등 사실 누가 작성했는지를 foreign key로  가지고 있기 때문에 createdBy, updatedBy와 중복된 값을  가지고 있을 확률이 높습니다.

이렇게 생각을 했을때 unique값인 nickname으로 관리하는게 더욱 낫다는 생각이 드네요 ㅎㅎ🤔

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/member/repository/MemberRepository.java:
```diff
@@ -0,0 +1,47 @@
+	 * 메서드의 파라미터를 쿼리 매개변수와 매핑하기 위해 사용한다.
+	 * Query annotation을 사용하여 직접 JPQL 쿼리를 정의할 때 매개변수를 지정하는 용도로 주로 사용한다.
+	 * */
+	@Query("SELECT m FROM Member m WHERE m.loginId = :loginId AND m.isDeleted = false")
```
멘토링떄 말씀드렸던것처럼 쿼리 어노테이션을 없애는 방향으로 다시 고민해주시면 좋을 것 같습니다~

jongwanra replied:
넵! 적용완료했습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+	private final EntityManager em;
+
+	@Override
+	public void signUp(
+		String loginId,
+		String nickname,
+		String plainPassword,
+		String passwordConfirm
```
메소드 파라미터 개행의 기준이 있나요~?

jongwanra replied:
개행 기준 보다는 Naver Coding Convention에서 제공하는 한줄 최대 길이 120을 유지하려고만 생각했습니다.

하지만 그 기준에 부합하지 않은 개행 조건으로 보이네요.  따로 컨벤션에 맞도록 변경하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+@RequiredArgsConstructor
+@Service
+@Transactional
+@Slf4j
```
지난 리뷰와 같습니다.

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	@Override
+	public void withdrawMember(Member member) {
```
1. 탈퇴 후 로그아웃 처리를 시키는 이유가 있으신가요~? SRP 관점에서 다시 한번 고민해보시면 좋을 것 같습니다. 
2. EntityManager 를 직접사용하신 이유가 있으신가요? flush 와 clear 가 반드시 필요하나요?

jongwanra replied:
1. 탈퇴 후 로그아웃 처리를 시키는 이유가 있으신가요~? SRP 관점에서 다시 한번 고민해보시면 좋을 것 같습니다.
=> 탈퇴 이후에  더 이상 접근을 못하게 하는게 맞다는 판단하에 탈퇴 후 로그아웃 처리도 별도로 해주었습니다. 
조금 더 고민을 해보니, JD님께서 말씀하신 부분 처럼 `탈퇴하다`의 행위와 `로그아웃하다`의 행위는 별개의 역할(책임)이다는 부분을  알 수 있었습니다!

2.  EntityManager 를 직접사용하신 이유가 있으신가요? flush 와 clear 가 반드시 필요하나요?
=> 탈퇴행위에 대한 테스트를 진행하면서 계속 실패 했던 이유가 쓰기 지연으로 인한 이슈라고 생각해서 넣었습니다!  결국엔 HttpSession을 제대로 Mocking해서 넣어주지 않았던 부분이였던 것 같습니다.😅

단순히 탈퇴행위만으로도 탈퇴한 회원은 접근이 불가능하다는 사실을 알 수 있었습니다. 반영해두겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/domain/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	@Override
+	public boolean isDuplicatedMemberByLoginId(String loginId) {
```
사용하는곳이 있지 않은 메소드인데 만드신 이유가 있으신가요? 
https://github.com/f-lab-edu/outfit-of-the-day/pull/6/files#diff-889efa49fd780db19e093c45e8a8a42d8533e1b5a20d1d9bb30922f638b517adR58 도 동일합니다. 

정작  signUp 메소드 에서는 `findByMemberByLoginIdOrNickname` 를 호출하고 있는데, 
필요한 메소드만 사용하고, 일관성있게 로직을 정리하면 좋을 것 같습니다.

jongwanra replied:
넵 정리하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,66 @@
+		@RequestBody @Valid AuthLogInRequestDto requestDto
+	) {
+		authService.login(requestDto.getLoginId(), requestDto.getPassword());
+		return new ResponseEntity<>(ApiResponseWrapper.of(SuccessCode.SUCCESS_POST), HttpStatus.OK);
```
SuccessCode 를 별도로 분리해서 관리하시는 이유가 있으신가요~? 
로그인,로그아웃의 경우 HttpStatus 는 204 도 고려해보아도 좋을 것 같습니다.

jongwanra replied:
뚜렷하게 성공에 대한 케이스만 모아서 관리하고 싶다는 생각에 적용한 코드였습니다!
하지만 생각할수록 SuccessCode를 관리하는 것 자체 보다는 기존에 존재하는 HttpStatus를 이용하는 방법이 훨씬더 접근하기 쉽고 유지 보수에도 용이하다고 생각이 들었습니다 ㅎㅎ 해당 부분 없애고 HttpStatus를 이용해서 접근하는 걸로 변경하겠습니다!

---

204 No Content

`성공했지만 response body에는 어떠한 데이터도 전달하지 않는 상태 코드` 

로그인과 로그아웃의 상황에서 알맞은 상태코드라고 생각이 듭니다! 수정 반영하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/common/config/SecurityConfig.java:
```diff
@@ -0,0 +1,39 @@
+	 * */
+	@Bean
+	public PasswordEncoder passwordEncoder() {
+		return new BCryptPasswordEncoder(8);
```
BCryptPasswordEncoder 동작방식에 대해서도 멘토링시간에 설명해주시면 좋을 것 같습니다.

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/auth/dto/AuthLogInRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.Getter;
+
+@Getter
+public class AuthLogInRequestDto {
```
Login vs LogIn 일관성있게 용어를 사용해주시면 좋을 것 같습니다.

jongwanra replied:
넵! `Login` 으로 일관되게 통일하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/common/response/ApiErrorResponse.java:
```diff
@@ -0,0 +1,31 @@
+
+@Getter
+public final class ApiErrorResponse {
+	private final Boolean isSuccess;
```
isSuccess 는 항상 false 가 아닌가요~?  그리고 boolean vs Boolean 선택의 기준은 무엇인가요?

jongwanra replied:
isSuccess 는 항상 false 가 아닌가요~?
=>  isSuccess는 항상 false인 부분 수정해서 반영하겠습니다!

boolean vs Boolean 선택의 기준은 무엇인가요?
=>  `Database Table과 매핑되는 Entity 객체에서는 Wrapper Class를, 그 외의 클래스에서는 primitive type을 가져가자` 의 기준으로 생각했습니다! 제대로 반영이 안되어 있는 것 같아서 다시 한 번 체크하겠습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/common/response/ApiResponseWrapper.java:
```diff
@@ -0,0 +1,31 @@
+@Getter
+public final class ApiResponseWrapper<T> {
+
+	private final boolean isSuccess;
```
isSuccess 가 항상 true 일 것 같습니다.

jongwanra replied:
success는 내부적으로  http 상태코드 400을 기준으로 성공과 실패를 구분할 수 있도록 변경했습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/common/response/ApiResponseWrapper.java:
```diff
@@ -0,0 +1,31 @@
+		return new ApiResponseWrapper<>(data, status, message);
+	}
+
+	public static <T> ApiResponseWrapper of(T data) {
```
of 보다는 success 나 ok 가 더 명확한 네이밍일 것 같습니다,

jongwanra replied:
위에 ApiResponseWrapper를 ApiResponse로 전체 변경하고,  
ApiResponse에 대해서 ok, badRequest, created 등의  method로 접근 가능하게 변경했습니다!
초기에는 응답에 대해서 성공과 실패를 나눠서 관리했다면, 현재 ApiResponse로 통일해서 관리할 수 있도록 변경했습니다 ㅎㅎ

예) ApIResponse.ok()

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/common/response/ErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+import lombok.Getter;
+
+@Getter
+public enum ErrorCode {
```
SuccessCode 와 하나로 합쳐서 관리를 해도 좋을 것 같습니다. 
(Success 에 대한 Code 가 근데 반드시 필요할까요~?)

jongwanra replied:
SuccessCode는 따로  ResponseApi Class에서  관리가 가능하고 불필요한 enum class로 저 역시 생각이 들었습니다!

해당 SuccessCode.java 파일 삭제 진행 이후에 ResponseApi.ok() method와 같이 사용할 수 있도록 반영해두었습니다

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/member/controller/MemberController.java:
```diff
@@ -0,0 +1,76 @@
+	 * 해당 annotation 내부에는 @RequestMapping(method = RequestMethod.GET)이 지정되어 있다.
+	 * Spring MVC에서 특정 GET 요청 URL을 처리하도록 매핑할 때 사용한다.
+	 * */
+	@GetMapping("/myself")
```
말씀드렸던것처럼 RestfulAPI 에 맞게 API URI 를 설계해주시면 좋을 것 같습니다.

jongwanra replied:
넵!

f-lab-jd commented on src/main/java/study/outfitoftheday/core/web/member/controller/MemberController.java:
```diff
@@ -0,0 +1,76 @@
+	public ResponseEntity<ApiResponseWrapper<String>> memberSignUp(
+		@RequestBody @Valid MemberSignUpRequestDto requestDto) {
+
+		memberService.signUp(requestDto.getLoginId(), requestDto.getNickname(), requestDto.getPassword(),
```
dto 를 그대로 전달하지 않고 여러개의 파라미터를 전달하시는 이유가 있으신가요~?

jongwanra replied:
Controller와 Service 레이어 간의 **의존 관계를 최대한 약하게 하고자** 따로 여러 개의 파라미터로 분리해서 적용했습니다!
강한 의존관계를 가지고 있으면 추후에 유지보수를 진행할 때, 더 많은 리소스를 소모할 것이라는 생각했습니다.

JD님께서 리뷰해주신 내용을 보고 조금 더 생각을 해보게 되었는데요,
지금은 제가 4가지의 파라미터 밖에 전달하지 않고 있지만,  이게 실무에서는 10개 이상이되면 어떨까? 하는 끔찍한 상상도 해보게 되었던 것 같아요. 이 방식에는 더 한계가 있을 것 같다고 생각이 들었습니다! 🥲

검색을 통해서 의존관계를 최소화 하면서 좋은 방법은 없는지 검색해보니

1. toEntity() method를 활용해서 Service의 파라미터로  Entity로 변환해서 전달하는 방법
2. Mapper Class를 활용해서 RequestDto와 Service단에서 Parameter로 활용할 Dto로 매핑해서 전달하는 방법


위 두가지가 떠올랐는데요,

첫 번째 케이스인 dto에서 따로 toEntity() method를 활용해서 만들어 준다면,
모든 로직에 대해서 일관되게 만들 수 없기도 하고 도메인 자체를 외부에서 주입하는 것과 핵심 비즈니스 로직을 외부에서 생성해서 주입을 한다는 것도 판단이 안서는 부분이였습니다.

두 번째 케이스인 Mapper Class를 활용해서 Service 계층에 전달할 parameter를 Mapper를 통해서  전달하면 어떨까에 대해선 
조금은 긍정적으로 받아들여졌습니다. 
따로 Controller, Service Layer에 의존 관계를 최소화 할 수 있고 비즈니스 로직을 외부에서 주입하지 않아도 된다는 점 때문이었습니다. 

하지만 이렇게 Mapper를 통해서 만들다보면 Dto를 작업하는 시간을 더욱 많이 할애할 것이라는 생각도 들었는데요,
Convention을 통해서 **Parameter 갯수가  4개 이상이라면 Mapper Class를 통해서 관리하면 어떨까 생각데**이 들었는데 

이 방향성에 대해서는 어떻게 생각하시는지 여쭤보고 싶습니다!

f-lab-jd replied:
먼저 1,2 두개 다 좋은 방법이고 잘 정리해주셨습니다. 👍  2번을 하게된다면 mapstruct 라는 라이브러리를 추천합니다. 
1,2 를 수행하는 곳이 Service 가 아니라 Controller 인 이유가 궁금합니다!
Service 와 Controller 각각의 역할이 무엇이라고 생각하시나요~?

jongwanra replied:
`mapstruct` 키워드 감사합니다ㅎㅎ👍🏻

Controller와 Service의 역할은 아래와 같이 이해했습니다!

* Controller는 Client로 부터의 요청에 대한 응답을 전달하는 역할을 수행합니다. 
* Service는 Repository와 상호작용하여 비즈니스 로직을 처리하는 역할을 수행합니다.

여기에 더해 Controller의 역할은 Client로 부터 전달받은 parameter의 값이 유효한지,  해당 parameter를 service에  전달하는 역할도 포함한다고 생각했습니다. 때문에 Controller에서 처리하는게 맞다고 판단했습니다!

f-lab-jd commented on src/main/resources/data/schema-dev.sql:
```diff
@@ -0,0 +1,116 @@
+    `is_deleted`        boolean            NOT NULL DEFAULT false
+);
+
+-- CREATE TABLE `comment`
```
테이블을 주석처리한 이유가 무엇인가요~? 
주석에는 주석의 이유가 담겨있으면 더 좋습니다.

jongwanra replied:
주석을 했을 때 **주석의 이유**를 담는 습관을 가지겠습니다!
Table에 주석처리를 해놨던 이유는 단순히 현재 사용하는 Entity 객체가 Member 밖에 없기 때문에 해두었습니다🤔
다시 한 번 생각해봤을 때, 주석을 할 필요가 없다는 생각이 들어서 주석을 다시 풀어서 적용시켜두겠습니다!

f-lab-jd commented on src/test/java/study/outfitoftheday/core/domain/member/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,121 @@
+ * default: true, method나 class에 지정 가능하다.
+ * */
+@SpringBootTest
+@Transactional
+@Rollback
```
테스트 코드가 서로 영향을 주지 않도록 만드는 것이 좋습니다.
`@Transactional` 과 `@Rollback`을 사용하여 테스트 간의 상태를 독립적으로 유지하고 있는데, 두개 다 필요할까요?

jongwanra replied:
Transactional 어노테이션은 테스트 환경에서 자동으로 rollback 처리까지 해주기 때문에,  Transactional 어노테이션 만으로 충분합니다.

해당 부분 반영하겠습니다!

f-lab-jd commented on src/test/java/study/outfitoftheday/core/domain/member/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,121 @@
+ * DB Transaction을 Rollback할지 여부를지정하는 annotation이다.
+ * default: true, method나 class에 지정 가능하다.
+ * */
+@SpringBootTest
```
실제 서비스로직만 테스트하려면 어떻게해야할까요?

jongwanra replied:
Mockito Framework를 이용해서  호출이 필요한 외부 모듈에 대한 모킹 처리를 통해 테스트를 진행하는 방법이 있습니다! ㅎㅎ

JD님 혹시 Service로직에 대해서 Mocking 테스트와 통합 테스트 중에 어떤 걸 가져가는 쪽을 선호하시는지 알 수 있을까요?

장/단점에 대해서 고려해 봤을 때, 테스트가 늘어날수록  비용과 테스트 수행 시간은 증가하는 상황에서 Mocking 테스트의 경우에는 통합 테스트에 비해 수행시간과 비용 측면에서 우위를 점하고 있습니다.  하지만 Mocking 테스트를 수행할 경우 객체들간의 협력의 관점에서 테스트를 수행해보지 않았기 때문에 확신할 수 없는 테스트이다 라는 생각도 듭니다!

f-lab-jd commented on src/test/java/study/outfitoftheday/core/web/member/controller/MemberControllerTest.java:
```diff
@@ -0,0 +1,154 @@
+@SpringBootTest
+@AutoConfigureMockMvc
+@AutoConfigureRestDocs
+@Rollback
+@Transactional
```
동일한 리뷰입니다.

f-lab-jd reviewed (changes requested):
고생하셨습니다. 리뷰 드린 내용 확인 부탁드립니다!

jongwanra committed:
[None] chore: System.out.println() method 제거

jongwanra committed:
[None] refactor: 매직 스트링 상수로 변경

jongwanra committed:
[None] refactor: auditorProvider() method HttpSession null 체크 제거

jongwanra committed:
[None] refactor: MemberController Rest Api에 맞게 URI 수정 및 테스트 코드 변경

jongwanra committed:
[None] refactor: MemberRepository Query Annotation 제거

jongwanra committed:
[None] refactor: schema.sql 주석 제거

jongwanra committed:
[None] refactor: createdBy, updatedBy 관련 초기값 설정

member를 처음 생성할 경우,
createdBy, updatedBy에 member_id값을 매핑하기에 어려움이 있다.

jongwanra committed:
[None] refactor: spring-security 제거 및 PasswordEncoder 변경

jongwanra committed:
[None] refactor: testa환경에서 lombok 제거

jongwanra committed:
[None] chore: application.yml file .gitignore에서 분리

jongwanra committed:
[None] chore: application.yml file .gitignore에서 분리

jongwanra committed:
[None] chore: ci.yml에서 mysql set up, application.yml 복사 작업 제외, gradlew test 작업 제외

jongwanra committed:
[None] chore: application.yml h2 database in-memory 방식으로 변경

jongwanra committed:
[None] chore: ci.yml에서 permissions.contents 옵션 삭제

jongwanra committed:
[None] refactor: withdrawMember method authService.logout method() 제거

jongwanra committed:
[None] refactor: Entity 객체를 제외한 Wrapper Class 기본 타입으로 변경 / ApiResponseWrapper 변경

jongwanra committed:
[None] refactor: Entity 객체를 제외한 Wrapper Class 기본 타입으로 변경 / ApiResponseWrapper 변경

jongwanra committed:
[None] refactor: @Slf4j 어노테이션 사용하지 않는 곳 제거

jongwanra committed:
[None] refactor: Login으로 용어 일관성 있게 통일

jongwanra committed:
[None] refactor: ApiResponse 구조 전체 변경

jongwanra committed:
[None] refactor: login/logout 상태 코드 204 No Content로 변경

jongwanra committed:
[None] refactor: 불필요한 예외 위임 삭제

jongwanra committed:
[None] refactor: LoginMemberArgumentResolver 에서 resolverArgument() method 반환 타입 Member로 변경

jongwanra committed:
[None] refactor: AuthService, MemberService Interface 삭제

jongwanra committed:
[None] refactor: MemberService.signUp() method paramter 개행된 부분 컨벤션에 맞게 변경

jongwanra committed:
[None] refactor: MemberRepository에서 사용하지 않는 method 삭제

jongwanra committed:
[None] refactor: AuthControllerTest authService mocking으로 구현 변경

jongwanra committed:
[None] refactor: Test 환경에서 @Rollback 어노테이션 제거

jongwanra committed:
[None] refactor: ServiceException으로 확장한 Custom Exception을 관리하는 방법으로 전체 Exception Handling 구성 변경

jongwanra committed:
[None] refactor: AuthService Transactional 어노테이션 제거

jongwanra committed:
[None] refactor: 전체 디렉토리 구조 변경

1. common directory -> global directory
2. BaseEntity -> global directory로 옮김
3. Controller directory 하위에 request, response directory 생성
4. global directory를 제외한 하위 common directory 전부 삭제
5. request, response의 prefix 'Dto' 제거

jongwanra committed:
[None] refactor: UriPrefix enum class에서 generic class로 변경 및 전체 적용

jongwanra committed:
[None] refactor: createdBy, updatedBy nickname으로 변경

jongwanra committed:
[None] refactor: restdocs 관련 build.gradle 파일 변경, 주석 추가

jongwanra committed:
[None] refactor: UriPrefix final class로 변경

jongwanra committed:
[None] refactor: RequestDto Service Parameter로 전달

jongwanra committed:
[None] refactor: SignUp 수행 시, passwordConfirm 관련 로직 제거

회원가입 시, 비밀번호 체크 로직은 Client쪽에서 해결하고 넘어오기 떄문에 불필요한 field로 판단.

jongwanra committed:
[None] refactor: ErrorCode에서 Exception 메시지 관리하는 로직 제거 및 별도로 Exception 메시지를 외부에서 주입하는 방식으로 변경

jongwanra committed:
[None] chore: develop 브랜치 push 시 동작하는 CI 전략으로 변경

jongwanra committed:
[#7] feat: build.gradle에 jacoco setting 추가

jongwanra committed:
[#7] test: AuthServiceTest 통합 테스트 코드 작성

jongwanra committed:
[#7] fix: build.gradle builddir deprecated issue 해결

jongwanra committed:
[#7] test: jacoco coverage 60%에 맞춰서 추가 테스트 코드 작성 및 Auth 관련 인증 실패 시 401, 인가 실패 시 403 로직 추가

jongwanra committed:
[None] chore: System.out.println() 제거

jongwanra committed:
[#7] feat: Pull Request시, CI 동작 확인하기

jongwanra committed:
[#7] feat: Pull Request시, CI 동작 확인하기 -2

jongwanra committed:
[#7] feat: Pull Request에서 Coverage 보이게 하는 CI 전략 추가

jongwanra committed:
[#7] feat: Pull Request에서 Coverage 보이게 하는 CI 전략 추가 2 - path 수정

jongwanra committed:
[#7] feat: Pull Request에서 Coverage 보이게 하는 CI 전략 추가 3 - 불필요한 옵션 제거

jongwanra committed:
[#7] feat: Pull Request에서 Coverage 보이게 하는 CI 전략 추가 4 - 불필요한 옵션 제거

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 1

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 2

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 3

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 4

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 5

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 6

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 7

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 8

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 9

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 10

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 11

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 12

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 13

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 14

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 15

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 16

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 17

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 18

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 19

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 20

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 21

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 22

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 23

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 24

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 25

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 26

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 27

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 28

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 29

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 30

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 31

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 32

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 33

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 34

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 35

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 36

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 37

jongwanra committed:
[#8] feat: GihubPage CI 전략 테스트 중 - 38

jongwanra committed:
[#8] feat: GihubPage에 rest docs 배포 관련 작업 ci-on-push.yml에 반영 완료

jongwanra committed:
Merge pull request #10 from f-lab-edu/feat/jongwan/github-page

[#8] feat: Github Page에 Spring Rest Docs 업로드 자동화

jongwanra committed:
[None] chore: GithubPage를 임시로 배포할 수 있도록 ci-on-push에 feat/jongwan/jacoco 브랜치 포함

jongwanra committed:
[#11] feature: logback 관련 초기 셋팅 및 appender 분리 작업

jongwanra committed:
[#11] feature: logback 관련 초기 셋팅 작업 추가

jongwanra committed:
[#11] feature: logback 운영 환경별 로그 전략 추가

jongwanra committed:
[#11] feature: logback 관련 테스트 코드 제거

f-lab-jd commented on src/main/java/study/outfitoftheday/global/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	@ExceptionHandler({ServiceException.class})
+	@ResponseStatus(HttpStatus.BAD_REQUEST)
```
Service Exception 을 상속받는 모든 예외가 발생한 경우 `BAD_REQUEST`  를 응답하도록 하신 이유가 있으신가요?

jongwanra replied:
Service Layer에서 발생하는 예외가 전부 400일 거라고 판단했습니다.!

JD님께서 주신 질문에 대해서  곰곰히 생각해 보고 찾아보면서 Service Layer에서 발생하는 예외를 전부 400으로 판단하면 안되겠구나 생각이 들었습니다. 

* 조회에 실패할 경우 발생하는 상태코드: 404
* Client가 너무 많은 요청을 할 경우 발생하는 상태코드: 429 

등 상태 코드에 맞는 Exception을 만들고 해당 Exception을 상속 받아서 상태를 상황에 알맞게 가져갈 수 있도록 다시 구성해 보겠습니다🙏

f-lab-jd commented on src/main/java/study/outfitoftheday/global/interceptor/RequiredAuthInterceptor.java:
```diff
@@ -0,0 +1,39 @@
+	private final AuthService authService;
+
+	@Override
+	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
```
`LoginMemberArgumentResolver` 클래스와 동일한 역할을 중복해서 하는 것 아닌가요? 
`ArgumentResolver vs Interceptor` 선택의 기준이 무엇이라고 생각하시나요?

jongwanra replied:
* *LoginMemberArgumentResolver 클래스와 동일한 역할을 중복해서 하는 것 아닌가요?*

`RequiredAuthInterceptor`를 통해서는  접근할 수 있는지 `인가`의 역할을  가지며 추후에 admin도 포함할 것을 고려해 추가했습니다.
`LoginMemberArgumentResolver`를 통해서는 'Login한 Member 조회'  역할을 위해서 추가했습니다!

* *ArgumentResolver vs Interceptor 선택의 기준이 무엇이라고 생각하시나요?*

ArgumentResolver는 controller method 실행 전 parameter에 값을 변환할 때 사용하며, 
Interceptor는 controller 호출 전/후로 로직을 처리할 때 사용합니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/global/response/ApiResponse.java:
```diff
@@ -0,0 +1,58 @@
+		return of(HttpStatus.BAD_REQUEST, message);
+	}
+	
+	private boolean isSuccess(int code) {
```
1xx~3xx 전부 성공이라고 처리하시는 이유가 있으신가요~?

jongwanra replied:
100번대와 300번대 상태 코드에 대해서 직접적으로 성공 혹은 실패라고 하기에 어렵다고 생각했습니다.! 
100번대는 추가적인 정보와 관련된 상태 코드 이며, 300번대는 리다이렉션과 관련된 상태 코드를 나타냅니다. 

100번대와 300번대 둘다 모호하여 더 간결한 코드로 접근했습니다..! 
null값을 허용하는 방향으로 접근하는 것은 어떨까요?!

f-lab-jd commented on src/main/java/study/outfitoftheday/web/member/controller/MemberController.java:
```diff
@@ -0,0 +1,73 @@
+	 * Spring MVC에서 특정 GET 요청 URL을 처리하도록 매핑할 때 사용한다.
+	 * */
+	@GetMapping("/{loginId}")
+	public ApiResponse<MemberGetByLoginIdResponse> memberGetByLoginId(
```
메소드명은 동사형으로 짓는 것이 가독성이 좋습니다.

jongwanra replied:
넵! 메서드를 전부 동사형으로 변경 진행 후 적용했습니다!

f-lab-jd commented on src/main/java/study/outfitoftheday/web/member/controller/MemberController.java:
```diff
@@ -0,0 +1,73 @@
+	 * 해당 annotation 내부에는 @RequestMapping(method = RequestMethod.GET)이 지정되어 있다.
+	 * Spring MVC에서 특정 GET 요청 URL을 처리하도록 매핑할 때 사용한다.
+	 * */
+	@GetMapping("/{loginId}")
```
`members` 에 대해 처리하는 것이니 loginId 보다는 memberId 가 맞을 것 같습니다.

jongwanra replied:
넵! memberId로 반영했고 관련 테스트 코드 추가했습니다!

f-lab-jd reviewed (approved):
고생하셨습니다!

jongwanra committed:
[#11] refactor: @AfterEach annotation -> @BeforeEach annotation으로 변경 및 test class에 @Transactional annotation제거

jongwanra committed:
[#11] refactor: logback appender level별 분리 제거

jongwanra committed:
[#11] refactor: logback profile별 파일 분리 및 공통 property 분리

jongwanra committed:
[#6] refactor: 메서드 동사형으로 변경 및 주어/목적어 제거

jongwanra committed:
[#6] refactor: GlobalExceptionHandler ServiceException 이 아닌 상태별 Exception으로 변경

jongwanra committed:
[#6] refactor: MemberController 조회 시, loginId -> memberId로 변경 및 MemberService 테스트 코드 추가

jongwanra committed:
[#11] chore: feat/jongwan/sign-up -> feat/jongwan/logging 머지 충돌 해결

jongwanra committed:
Merge pull request #12 from f-lab-edu/feat/jongwan/logging

[#11] feat: 로깅툴 logback으로 적용하기

jongwanra committed:
[None] chore: jacocoTestCoverageVerification
exception, response directory excludes에 추가

f-lab-jd commented:
리뷰 이후에 커밋이 추가 되었는데, 가급적이면 리뷰 후 (간단하게 반영할거리가 있다면 반영 후) 머지하면 좋을 것 같습니다 ㅎㅎ

jongwanra committed:
[#14] feat: spring boot actuator 기본 설정

jongwanra commented:
*리뷰 이후에 커밋이 추가 되었는데, 가급적이면 리뷰 후 (간단하게 반영할거리가 있다면 반영 후) 머지하면 좋을 것 같습니다 ㅎㅎ*
-> 넵!  전체 마무리 되고 PR 적용 직전에 머지하는 습관 가지겠습니다 🙏

jongwanra committed:
Merge pull request #15 from f-lab-edu/feat/jongwan/actuator

[#14] feat: spring boot actuator 기본 설정
